File: linked_list/reverse_linked_list.rs
// Solution for: https://leetcode.com/problems/reverse-linked-list/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut prev = None;
    let mut next = head;
    while next.is_some() {
        let mut cur = next.unwrap(); // Box<ListNode>
        next = cur.next.take(); // next = next.next ; become next
        cur.next = prev;
        prev = Some(cur);
    }
    prev
}

File: linked_list/reverse_linked_list.cpp
// Solution for: https://leetcode.com/problems/reorder-list/submissions/

#include "linked_list/defs.hpp"

ListNode *reverseList(ListNode *head)
{
    ListNode *prev = nullptr;
    ListNode *curr = head;
    ListNode *next = nullptr;
    while (curr != nullptr)
    {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}